bpf/aya-bpf/src/maps/ring_buf.rs

```diff
@@ -0,0 +1,202 @@
+use core::{
+    cell::UnsafeCell,
+    mem,
+    mem::MaybeUninit,
+    ops::{Deref, DerefMut},
+};
+
+use crate::{
+    bindings::{bpf_map_def, bpf_map_type::BPF_MAP_TYPE_RINGBUF},
+    helpers::{
+        bpf_ringbuf_discard, bpf_ringbuf_output, bpf_ringbuf_query, bpf_ringbuf_reserve,
+        bpf_ringbuf_submit,
+    },
+    maps::PinningType,
+};
+
+#[repr(transparent)]
+pub struct RingBuf {
+    def: UnsafeCell<bpf_map_def>,
+}
+
+unsafe impl Sync for RingBuf {}
+
+/// A ring buffer entry, returned from [`RingBuf::reserve`].
+///
+/// You must [`submit`] or [`discard`] this entry before this gets dropped.
+///
+/// [`submit`]: RingBufEntry::submit
+/// [`discard`]: RingBufEntry::discard
+#[must_use = "BPF verifier requires ring buffer entries to be either submitted or discarded"]
+pub struct RingBufEntry<T: 'static>(&'static mut MaybeUninit<T>);
+
+impl<T> Deref for RingBufEntry<T> {
+    type Target = MaybeUninit<T>;
+
+    fn deref(&self) -> &Self::Target {
+        self.0
+    }
+}
+
+impl<T> DerefMut for RingBufEntry<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.0
+    }
+}
+
+impl<T> RingBufEntry<T> {
+    /// Discard this ring buffer entry. The entry will be skipped by the userspace reader.
+    pub fn discard(self, flags: u64) {
+        unsafe { bpf_ringbuf_discard(self.0.as_mut_ptr() as *mut _, flags) };
+    }
+
+    /// Commit this ring buffer entry. The entry will be made visible to the userspace reader.
+    pub fn submit(self, flags: u64) {
+        unsafe { bpf_ringbuf_submit(self.0.as_mut_ptr() as *mut _, flags) };
+    }
+}
+
+impl RingBuf {
+    /// Declare a BPF ring buffer.
+    ///
+    /// `byte_size` should be a power-of-2 multiple of the page size. If it is not, it will
+    /// be coerced to the next largest valid size.
+    pub const fn with_byte_size(byte_size: u32, flags: u32) -> Self {
+        Self::new(byte_size, flags, PinningType::None)
+    }
+
+    /// Declare a pinned BPF ring buffer.
+    ///
+    /// `byte_size` should be a power-of-2 multiple of the page size. If it is not, it will
+    /// be coerced to the next largest valid size.
+    pub const fn pinned(byte_size: u32, flags: u32) -> Self {
+        Self::new(byte_size, flags, PinningType::ByName)
+    }
+
+    const fn new(byte_size: u32, flags: u32, pinning_type: PinningType) -> Self {
+        Self {
+            def: UnsafeCell::new(bpf_map_def {
+                type_: BPF_MAP_TYPE_RINGBUF,
+                key_size: 0,
+                value_size: 0,
+                max_entries: adjust_to_page_size(byte_size),
```

Comment:
I wouldn't try and do this on the bpf-side of the prog.
It can be done in userspace when we create the map.

Reply:
That makes sense. Also that's how the libbpf code works. Thanks for catching this. Done.